kame/kame/dhcp6/dhcp6sctl.8
.\"	$KAME: dhcp6sctl.8,v 1.2 2004/07/29 23:39:26 jinmei Exp $
.\"
.\" Copyright (C) 2004 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 19, 2004
.Dt DHCP6SCTL 8
.Os KAME
.Sh NAME
.Nm dhcp6sctl
.Nd DHCPv6 server control utility
.\"
.Sh SYNOPSIS
.Nm
.Op Fl k Ar keyfile
.Op Fl p Ar port
.Op Fl s Ar server_address
.Ic command
.\"
.Sh DESCRIPTION
.Nm
controls the operation of
.Nm dhcp6s ,
a DHCPv6 server.
.Pp
.Nm
communicates with the DHCPv6 server over a TCP connection,
sending commands authenticated with digital signatures.
Currently,
the only supported authentication algorithm is
HMAC-MD5,
which uses a shared secret on each end of the connection.
.Pp
Command line options are as below:
.Bl -tag -width indent
.\"
.It Fl k Ar keyfile
Use
.Ar keyfile
to provide the shared secret to communicate with the server.
The default file name used when unspecified is
.Pa /usr/local/v6/etc/dhcp6sctlkey .
.It Fl p Ar port
Specify
.Ar port
as the listening port of the server.
The default port number used when unspecified is 5546.
.It Fl s Ar server_address
Specify
.Ar server_address
as the listening address of the server.
The default address used when unspecified is ::1.
.El
.\"
.Sh KEY FILE
Since the operation available with the
.Nm
command is powerful,
the communication between the command and
.Nm dhcp6s
must be authenticated.
The supported algorithm for authentication is HMAC-MD5,
which requires a shared secret,
and the secret is stored in the key file.
The key file must consist of a single line, in which the secret value
is written in the form of BASE-64 encoding.
.\"
.Sh COMMANDS
Each
.Ic command
specifies a single control operation.
Supported commands are as follows:
.Pp
.Bl -tag -width Ds -compact
.It Xo
.Ic reload
.Xc
This command specifies the server to reload the configuration file.
Existing bindings, if any, are intact.
.It Xo
.Ic remove Ar arguments
.Xc
This command specifies the server to remove a run-time object
specified by
.Ar arguments
from the server.
Currently only possible object is one particular IA_PD binding,
which is specified as
.Ic IA Ic IA_PD Ar IAID Ar DUID
where
.Ar IAID
is a decimal number specifying the ID of identity association (IA),
and
.Ar DUID
is a DHCP Unique Identifier of the binding.
The format of
.Ar DUID
is the same as that specified in
.Xr dhcp6s.conf 5 .
.El
.\"
.Sh FILES
.Bl -tag -width /usr/local/v6/etc/dhcp6sctlkey -compact
.It Pa /usr/local/v6/etc/dhcp6sctlkey
is the default key file.
.El
.\"
.Sh SEE ALSO
.Xr dhcp6s.conf 5
.Xr dhcp6s 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME IPv6 protocol stack kit.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/dhcp6/dhcp6sctl.c
/*	$KAME: dhcp6sctl.c,v 1.10 2004/08/01 07:34:34 suz Exp $	*/

/*
 * Copyright (C) 2004 WIDE Project.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include <sys/types.h>
#include <sys/socket.h>
#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <netinet/in.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <netdb.h>
#include <err.h>

#include <control.h>
#include <auth.h>
#include <base64.h>

#define MD5_DIGESTLENGTH 16
#define DEFAULT_KEYFILE SYSCONFDIR "/dhcp6sctlkey"

static char *ctladdr = DEFAULT_CONTROL_ADDR;
static char *ctlport = DEFAULT_CONTROL_PORT;

static inline int put16 __P((char **, int *, u_int16_t));
static inline int put32 __P((char **, int *, u_int32_t));
static inline int putval __P((char **, int *, void *, size_t));

static int setup_auth __P((char *, struct keyinfo *, int *));
static int make_command __P((int, char **, char **, size_t *,
    struct keyinfo *, int));
static int make_remove_command __P((int, char **, char **, int *));
static int make_binding_object __P((int, char **, char **, int *));
static int make_ia_object __P((int, char **, char **, int *));
static int parse_duid __P((char *, int *, char **, int *));
static void usage __P((void));

int
main(argc, argv)
	int argc;
	char *argv[];
{
	int cc, ch, s, error, passed;
	char *cbuf;
	size_t clen;
	struct addrinfo hints, *res0, *res;
	int digestlen;
	char *keyfile = DEFAULT_KEYFILE;
	struct keyinfo key;

	while ((ch = getopt(argc, argv, "k:p:s:")) != -1) {
		switch (ch) {
		case 'k':
			keyfile = optarg;
			break;
		case 'p':
			ctlport = optarg;
			break;
		case 's':
			ctladdr = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage();

	memset(&key, 0, sizeof(key));
	digestlen = 0;
	if (setup_auth(keyfile, &key, &digestlen) != 0)
		errx(1, "failed to setup message authentication");

	if ((passed = make_command(argc, argv, &cbuf, &clen,
	    &key, digestlen)) < 0) {
		errx(1, "failed to make command buffer");
	}
	argc -= passed;
	argv += passed;
	if (argc != 0)
		warnx("redundant command argument after \"%s\"", argv[0]);

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_INET6;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_protocol = IPPROTO_TCP;
	error = getaddrinfo(ctladdr, ctlport, &hints, &res0);
	if (error != 0)
		errx(1, "getaddrinfo failed: %s", gai_strerror(error));

	s = -1;
	for (res = res0; res != NULL; res = res->ai_next) {
		s = socket(res->ai_family, res->ai_socktype,
		    res->ai_protocol);
		if (s < 0) {
			warn("socket");
			continue;
		}
		if (connect(s, res->ai_addr, res->ai_addrlen) < 0) {
			warn("connect");
			s = -1;
			continue;
		}
		break;
	}
	freeaddrinfo(res0);
	if (s < 0) {
		warnx("failed to connect to the server");
		exit(1);
	}

	cc = write(s, cbuf, clen);
	if (cc < 0)
		err(1, "write command");
	if (cc != clen)
		errx(1, "failed to send complete command");

	close(s);
	free(cbuf);

	exit(0);
}

static int
setup_auth(keyfile, key, digestlenp)
	char *keyfile;
	struct keyinfo *key;
	int *digestlenp;
{
	FILE *fp = NULL;
	char line[1024], secret[1024];
	int secretlen;

	key->secret = NULL;

	/* Currently, we only support HMAC-MD5 for authentication. */
	*digestlenp = MD5_DIGESTLENGTH;

	if ((fp = fopen(keyfile, "r")) == NULL) {
		warn("fopen: %s", keyfile);
		return (-1);
	}
	if (fgets(line, sizeof(line), fp) == NULL && ferror(fp)) {
		warn("fgets failed");
		goto fail;
	}
	if ((secretlen = base64_decodestring(line, secret, sizeof(secret)))
	    < 0) {
		warnx("failed to decode base64 string");
		goto fail;
	}
	if ((key->secret = malloc(secretlen)) == NULL) {
		warn("setup_auth: malloc failed");
		goto fail;
	}
	key->secretlen = (size_t)secretlen;
	memcpy(key->secret, secret, secretlen);

	fclose(fp);

	return (0);

  fail:
	if (fp != NULL)
		fclose(fp);
	if (key->secret != NULL)
		free(key->secret);
	return (-1);
}

static inline int
put16(bpp, lenp, val)
	char **bpp;
	int *lenp;
	u_int16_t val;
{
	char *bp = *bpp;
	int len = *lenp;

	if (len < sizeof(val))
		return (-1);

	val = htons(val);
	memcpy(bp, &val, sizeof(val));
	bp += sizeof(val);
	len -= sizeof(val);

	*bpp = bp;
	*lenp = len;

	return (0);
}

static inline int
put32(bpp, lenp, val)
	char **bpp;
	int *lenp;
	u_int32_t val;
{
	char *bp = *bpp;
	int len = *lenp;

	if (len < sizeof(val))
		return (-1);

	val = htonl(val);
	memcpy(bp, &val, sizeof(val));
	bp += sizeof(val);
	len -= sizeof(val);

	*bpp = bp;
	*lenp = len;

	return (0);
}

static inline int
putval(bpp, lenp, val, valsize)
	char **bpp;
	int *lenp;
	void *val;
	size_t valsize;
{
	char *bp = *bpp;
	int len = *lenp;

	if (len < valsize)
		return (-1);

	memcpy(bp, val, valsize);
	bp += valsize;
	len -= valsize;

	*bpp = bp;
	*lenp = len;

	return (0);
}

static int
make_command(argc, argv, bufp, lenp, key, authlen)
	int argc;
	char **argv, **bufp;
	size_t *lenp;
	struct keyinfo *key;
	int authlen;
{
	struct dhcp6ctl ctl;
	char commandbuf[4096];	/* XXX: ad-hoc value */
	char *bp, *buf, *mac;
	int buflen, len;
	int argc_passed = 0, passed;
	time_t now;

	if (argc == 0) {
		warnx("command is too short");
		return (-1);
	}

	bp = commandbuf + sizeof(ctl) + authlen;
	if (bp >= commandbuf + sizeof(commandbuf)) {
		warnx("make_command: local buffer is too short");
		return (-1);
	}
	buflen = sizeof(commandbuf) - sizeof(ctl);

	memset(&ctl, 0, sizeof(ctl));
	ctl.version = htons(DHCP6CTL_VERSION);

	if (strcmp(argv[0], "reload") == 0)
		ctl.command = htons(DHCP6CTL_COMMAND_RELOAD);
	else if (strcmp(argv[0], "remove") == 0) {
		if ((passed = make_remove_command(argc - 1, argv + 1,
		    &bp, &buflen)) < 0) {
			return (-1);
		}
		argc_passed += passed;
		ctl.command = htons(DHCP6CTL_COMMAND_REMOVE);
	} else {
		warnx("unknown command: %s", argv[0]);
		return (-1);
	}

	len = bp - commandbuf;
	ctl.len = htons(len - sizeof(ctl));

	if ((now = time(NULL)) < 0) {
		warn("failed to get current time");
		return (-1);
	}
	ctl.timestamp = htonl((u_int32_t)now);

	memcpy(commandbuf, &ctl, sizeof(ctl));

	mac = commandbuf + sizeof(ctl);
	memset(mac, 0, authlen);
	if (dhcp6_calc_mac(commandbuf, len, DHCP6CTL_AUTHPROTO_UNDEF,
	    DHCP6CTL_AUTHALG_HMACMD5, sizeof(ctl), key) != 0) {
		warnx("failed to calculate MAC");
		return (-1);
	}

	if ((buf = malloc(len)) == NULL) {
		warn("memory allocation failed");
		return (-1);
	}
	memcpy(buf, commandbuf, len);

	*lenp = len;
	*bufp = buf;

	argc_passed++;

	return (argc_passed);
}

static int
make_remove_command(argc, argv, bpp, lenp)
	int argc, *lenp;
	char **argv, **bpp;
{
	int argc_passed = 0, passed;

	if (argc == 0) {
		warnx("make_remove_command: command is too short");
		return (-1);
	}

	if (strcmp(argv[0], "binding") == 0) {
		if (put32(bpp, lenp, DHCP6CTL_BINDING))
			goto fail;
		if ((passed = make_binding_object(argc - 1, argv + 1,
		    bpp, lenp)) < 0) {
			return (-1);
		}
		argc_passed += passed;
	} else {
		warnx("remove target not supported: %s", argv[0]);
		return (-1);
	}

	argc_passed++;
	return (argc_passed);

  fail:
	warnx("make_remove_command failed");
	return (-1);
}

static int
make_binding_object(argc, argv, bpp, lenp)
	int argc, *lenp;
	char **argv, **bpp;
{
	int argc_passed = 0, passed;

	if (argc == 0) {
		/* or allow this as "all"? */
		warnx("make_binding_object: command is too short");
		return (-1);
	}

	if (strcmp(argv[0], "IA") == 0) {
		if (put32(bpp, lenp, DHCP6CTL_BINDING_IA))
			goto fail;
		if ((passed = make_ia_object(argc - 1, argv + 1,
		    bpp, lenp)) < 0) {
			return (-1);
		}
		argc_passed += passed;
	} else {
		warn("unknown binding type: %s", argv[0]);
		return (-1);
	}

	argc_passed++;
	return (argc_passed);

  fail:
	warnx("make_binding_object: failed");
	return (-1);
}

static int
make_ia_object(argc, argv, bpp, lenp)
	int argc, *lenp;
	char **argv, **bpp;
{
	struct dhcp6ctl_iaspec iaspec;
	int duidlen, dummylen = 0;
	int argc_passed = 0;
	char *dummy = NULL;

	if (argc < 3) {
		/*
		 * Right now, we require all three parameters of
		 * <IA type, IAID, DUID>.  This should be more flexible in
		 * the future.
		 */
		warnx("command is too short for an IA spec");
		return (-1);
	}
	argc_passed += 3;

	memset(&iaspec, 0, sizeof(iaspec));

	if (strcmp(argv[0], "IA_PD") == 0)
		iaspec.type = htonl(DHCP6CTL_IA_PD);
	else {
		warnx("IA type not supported: %s", argv[0]);
		return (-1);
	}

	iaspec.id = htonl((u_int32_t)strtol(argv[1], NULL, 10));

	if (parse_duid(argv[2], &duidlen, &dummy, &dummylen))
		goto fail;
	iaspec.duidlen = htonl(duidlen);

	if (putval(bpp, lenp, &iaspec, sizeof(iaspec)))
		goto fail;

	if (parse_duid(argv[2], &duidlen, bpp, lenp))
		goto fail;

	return (argc_passed);

  fail:
	warnx("make_ia_object: failed");
	return (-1);
}

static int
parse_duid(str, lenp, bufp, buflenp)
	char *str;
	int *lenp;
	char **bufp;
	int *buflenp;
{
	char *buf = *bufp;
	char *cp, *bp;
	int duidlen, slen, buflen;
	unsigned int x;

	/* calculate DUID len */
	slen = strlen(str);
	if (slen < 2)
		goto bad;
	duidlen = 1;
	slen -= 2;
	if ((slen % 3) != 0)
		goto bad;
	duidlen += (slen / 3);
	if (duidlen > 128) {
		warn("too long DUID (%d bytes)", duidlen);
		return (-1);
	}

	*lenp = duidlen;
	if (buf == NULL)
		return (0);

	buflen = *buflenp;
	if (buflen < duidlen)
		goto bad;

	for (cp = str, bp = buf; *cp != '\0';) {
		if (bp - buf >= buflen)
			goto bad;

		if (sscanf(cp, "%02x", &x) != 1)
			goto bad;
		*bp++ = x;
		cp += 2;

		switch (*cp) {
		case ':':
			cp++;
			break;
		case '\0':
			goto done;
		default:
			goto bad;
		}
	}
  done:
	*bufp = bp;
	return (0);

  bad:
	return (-1);
}

static void
usage()
{
	fprintf(stderr, "usage: dhcp6sctl [-k keyfile] [-p port] "
	    "[-s server_address] command...\n");

	exit(1);
}
